File: SpiralMatrixII.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        
        vector< vector <int> > ans(n, vector<int>(n,0));
        
        int val = 1;
        
        int t = 0, b = n-1, l = 0, r = n-1, d = 0;
        
        while(t <= b && l <= r){
            
            //vector<int> row;
            if(d == 0){
                
                for(int i = l; i <= r;i++){
                    ans[t][i] = val++;
                }
                t++;
            }
            else if(d == 1){
                
                for(int i = t; i <= b;i++){
                    ans[i][r] = val++;
                }
                r--;
            }
            else if(d == 2){
                for(int i = r; i >= l;i--){
                    ans[b][i] = val++;
                }
                b--;
            }
            else if(d == 3){
                for(int i = b; i >= t;i--){
                    ans[i][l] = val++;
                }
                l++;
            }
            
            //ans.push_back(row);
            
            d = (d+1)%4;
        }
        
        return ans;
        
    }
};

File: SpiralMatrixII.java
class Solution {
    public int[][] generateMatrix(int n) {
        
        int d = 0, top = 0, bottom = n - 1, left = 0, right = n - 1;

        int[][] mat = new int[n][n];
        int start = 1;
        while(left <= right && top <= bottom){
            switch(d){
                case 0:
                    for(int i = left;i <= right;i++)mat[top][i] = start++;
                    top++;
                    break;
                case 1: 
                    for(int i = top;i <= bottom;i++)mat[i][right] = start++;
                    right--;
                    break;
                case 2:
                    for(int i = right; i >= left;i--)mat[bottom][i] = start++;
                    bottom--;
                    break;
                case 3:
                    for(int i = bottom; i >= top;i--)mat[i][left] = start++;
                    left++;
                    break;
            }

            d = (d + 1) % 4;
        }

        return mat;
    }

}
